fedimint-client/src/db/event_log.rs

```diff
@@ -0,0 +1,460 @@
+impl<'tx, Cap> DBTransactionEventLogExt for DatabaseTransaction<'tx, Cap>
+where
+    Cap: Send,
+{
+    async fn log_event_raw(
+        &mut self,
+        log_ordering_wakeup_tx: watch::Sender<()>,
+        kind: EventKind,
+        module: Option<ModuleKind>,
+        payload: Vec<u8>,
+    ) {
+        let unordered_id = UnordedEventLogId::new();
+        trace!(target: LOG_CLIENT_EVENT_LOG, ?unordered_id, "New unordered event log event");
+
+        if self
+            .insert_entry(
+                &unordered_id,
+                &EventLogEntry {
+                    kind,
+                    module,
+                    ts_msecs: unordered_id.ts_msecs,
+                    payload,
+                },
+            )
+            .await
+            .is_some()
+        {
+            panic!("Trying to overwrite event in the client event log");
+        }
+        self.on_commit(move || {
+            let _ = log_ordering_wakeup_tx.send(());
+        });
+    }
+
+    async fn log_event<E>(&mut self, log_ordering_wakeup_tx: watch::Sender<()>, event: E)
+    where
+        E: Event + Send,
+    {
+        self.log_event_raw(
+            log_ordering_wakeup_tx,
+            E::KIND,
+            E::MODULE,
+            serde_json::to_vec(&event).expect("Serialization can't fail"),
+        )
+        .await;
+    }
+
+    async fn get_next_event_log_id(&mut self) -> EventLogId {
+        self.find_by_prefix_sorted_descending(&EventLogIdPrefixAll)
+            .await
+            .next()
+            .await
+            .map(|(k, _v)| k.next())
+            .unwrap_or_default()
+    }
+
+    async fn get_event_log(
+        &mut self,
+        pos: Option<EventLogId>,
+        limit: usize,
+    ) -> BTreeMap<EventLogId, (EventKind, Option<ModuleKind>, u64, serde_json::Value)> {
+        // TODO: This is innefficent for large logs, and there's no good operations to
+        // help? Do we just query one by one instead?
+        self.find_by_prefix(&EventLogIdPrefixAll)
+            .await
+            .filter(|(k, _)| future::ready(pos.unwrap_or_default() <= *k))
+            .take(limit)
+            .map(|(k, v)| {
+                (
+                    k,
+                    (
+                        v.kind,
+                        v.module,
+                        v.ts_msecs,
+                        serde_json::from_slice(&v.payload).unwrap_or_default(),
+                    ),
+                )
+            })
+            .collect()
+            .await
+    }
+}
+
+pub async fn run_event_log_ordering_task(
+    db: Database,
+    mut log_ordering_task_wakeup: watch::Receiver<()>,
+    log_event_added: watch::Sender<()>,
+) {
+    let mut next_entry_id = db
+        .begin_transaction_nc()
+        .await
+        .get_next_event_log_id()
+        .await;
+
+    loop {
+        let mut dbtx = db.begin_transaction().await;
+
+        let unordered_events = dbtx
+            .find_by_prefix(&UnorderedEventLogIdPrefixAll)
```

Comment:
Since `UnordedEventLogId` is mostly ordered, do we still want to do sorting here? Does `find_by_prefix` already give us implicit ordering?

Reply:
`find_by_prefix` sorts by key.


The most important part of it is that they will all get contiguous, numerical ID here, as they are being moved to the ordered final event log.

Reply:
Ahh gotcha, I stopped at the docs instead of reading the implementations :slightly_smiling_face: 

[nit] It might be worth updating `raw_find_by_prefix` docs to read "sorted ascending by key" instead of "No particular ordering is guaranteed."

https://github.com/fedimint/fedimint/blob/6e552f42657474325cc4b4980d21e577adbd5fd1/fedimint-core/src/db/mod.rs#L799-L801

Reply:
Oh... :D